src/posts/flox-smart/index.md

```diff
@@ -0,0 +1,215 @@
+
+## What is flox?
+
+[`flox` implements](https://flox.readthedocs.io/) grouped reductions for chunked array types like [cubed](https://cubed-dev.github.io/cubed/) and [dask](https://docs.dask.org/en/stable/array.html) using tree reductions.
+Tree reductions ([example](https://people.csail.mit.edu/xchen/gpu-programming/Lecture11-reduction.pdf)) are a parallel-friendly way of computing common reduction operations like `sum`, `mean` etc.
+Briefly, one computes the reduction for a subset of the array N chunks at a time in parallel, then combines those results together again N chunks at a time, until we have the final result.
+
+Without flox, Xarray effectively shuffles — sorts the data to extract all values in a single group — and then runs the reduction group-by-group.
+Depending on data layout or "chunking" this shuffle can be quite expensive.
+Here's a schematic of an array with 5 chunks arranged vertically, and each chunk has 10 elements each of which are colored by group.
+![shuffle](https://flox.readthedocs.io/en/latest/_images/new-split-apply-combine-annotated.svg)
+With flox installed, Xarray instead uses its parallel-friendly tree reduction.
+In many cases, this is a massive improvement.
+Notice how much cleaner the graph is in this image:
+![map-reduce](https://flox.readthedocs.io/en/latest/_images/new-map-reduce-reindex-True-annotated.svg)
+See our [previous blog post](https://xarray.dev/blog/flox) or the [docs](https://flox.readthedocs.io/en/latest/implementation.html#method-map-reduce) for more.
+
+## Tree reductions can be catastrophically bad
+
+Consider `ds.groupby("time.year").mean()`, or the equivalent `ds.resample(time="Y").mean()` for a 100 year long dataset of monthly averages with chunk size of **1** (or **4**) along the time dimension.
+This is a fairly common format for climate model output.
+The small chunk size along time is offset by much larger chunk sizes along the other dimensions — commonly horizontal space (`x, y` or `latitude, longitude`).
+
+A naive tree reduction would accumulate all averaged values into a single output chunk of size 100 — one value per year for 100 years.
+Depending on the chunking of the input dataset, this may overload the final worker's memory and fail catastrophically.
+More importantly, there is a lot of wasteful communication — computing on the last year of data is completely independent of computing on the first year of the data, and there is no reason the results for the two years need to reside in the same output chunk.
+This issue does _not_ arise for regular reductions where the final result depends on the values in all chunks, and all data along the reduced axes are reduced down to one final value.
+
+## Avoiding catastrophe
+
+Thus `flox` quickly grew two new modes of computing the groupby reduction.
```

Comment:
I would move this sentence down to after you have gone through your two realizations. Otherwise it's confusing when you jump into 

> First, `method="blockwise"`